plugin/pkg/admission/eventratelimit/admission.go

```diff
@@ -0,0 +1,270 @@
+	return &singleCache{
+		rateLimiter: rateLimiter,
+	}
+}
+
+func (c *singleCache) get(key interface{}) (flowcontrol.RateLimiter, bool) {
+	return c.rateLimiter, true
+}
+
+func (c *singleCache) add(key interface{}, rateLimiter flowcontrol.RateLimiter) {
+	// Do nothing as the rate limiter stored by this cache is constant
+}
+
+// lruCache is a least-recently-used cache
+type lruCache struct {
+	cache *lru.Cache
+}
+
+func newLRUCache(cacheSize int) *lruCache {
+	return &lruCache{
+		cache: lru.New(cacheSize),
+	}
+}
+
+func (c *lruCache) get(key interface{}) (flowcontrol.RateLimiter, bool) {
+	value, found := c.cache.Get(key)
+	if !found {
+		return nil, false
+	}
+	return value.(flowcontrol.RateLimiter), true
+}
+
+func (c *lruCache) add(key interface{}, rateLimiter flowcontrol.RateLimiter) {
+	c.cache.Add(key, rateLimiter)
+}
+
+// factory for creating a rate limiter
+type rateLimiterFactory func(qps float32, burst int) flowcontrol.RateLimiter
+
+// newEventRateLimit configures an admission controller that can enforce event rate limits
+func newEventRateLimit(config *eventratelimitapi.Configuration) (admission.Interface, error) {
+	rlf := func(qps float32, burst int) flowcontrol.RateLimiter {
+		return flowcontrol.NewTokenBucketRateLimiter(qps, burst)
+	}
+	return newEventRateLimitUsingRLF(config, rlf)
+}
+
+// newEventRateLimitWithClock configures an admission controller that can enforce event rate limits.
+// It uses a clock for testing purposes.
+func newEventRateLimitWithClock(config *eventratelimitapi.Configuration, clock flowcontrol.Clock) (admission.Interface, error) {
+	rlf := func(qps float32, burst int) flowcontrol.RateLimiter {
+		return flowcontrol.NewTokenBucketRateLimiterWithClock(qps, burst, clock)
+	}
+	return newEventRateLimitUsingRLF(config, rlf)
+}
+
+// newEventRateLimit configures an admission controller than can enforce event rate limits
+func newEventRateLimitUsingRLF(config *eventratelimitapi.Configuration, rlf rateLimiterFactory) (admission.Interface, error) {
+	eventRateLimitAdmission := &eventRateLimitAdmission{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}
+
+	for _, limitConfig := range config.Limits {
+		enforcer, err := newLimitEnforcer(limitConfig, rlf)
+		if err != nil {
+			return nil, err
+		}
+		eventRateLimitAdmission.limitEnforcers = append(eventRateLimitAdmission.limitEnforcers, enforcer)
+	}
+
+	return eventRateLimitAdmission, nil
+}
+
+// Admit makes admission decisions while enforcing event rate limits
+func (a *eventRateLimitAdmission) Admit(attr admission.Attributes) (err error) {
+	// ignore all operations that do not correspond to an Event kind
+	if attr.GetKind().GroupKind() != api.Kind("Event") {
+		return nil
+	}
+
+	a.Lock()
+	defer a.Unlock()
+
+	// give each limit enforcer a chance to reject the event
+	for _, enforcer := range a.limitEnforcers {
+		if err := enforcer.accept(attr); err != nil {
+			return err
```

Comment:
you can't short-circuit here.  You need to count against all the rate limits.

Reply:
Derek and I discussed this a while back, coming to the conclusion that it was adequate to short-circuit. The behavior is similar to having each limit type be a separate admission control. If an earlier admission control in the chain rejects a request, the later admission controls never see that request and do not have the opportunity to count the request against any limits. I don't think that it really matters too much in the grand scheme of things whether we count a rejected request against all limits or only a subset of limits. If you feel strongly that each rejected request must count against all limits, then I will make that change. Logically, it probably makes the most sense to have a rejected request count against none of the limits, as it was rejected. This is harder to implement, though. And, again, I don't think it matters too much in the overall flow of event requests.

Reply:
> If you feel strongly that each rejected request must count against all limits, then I will make that change.

I feel pretty strongly.  Either all or none please